internal/server/get_analysisrun_logs_v1alpha1.go

```diff
@@ -0,0 +1,393 @@
+package server
+
+import (
+	"bufio"
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+
+	"connectrpc.com/connect"
+	"github.com/hashicorp/go-cleanhttp"
+	"k8s.io/apimachinery/pkg/types"
+
+	svcv1alpha1 "github.com/akuity/kargo/api/service/v1alpha1"
+	rolloutsapi "github.com/akuity/kargo/api/stubs/rollouts/v1alpha1"
+	kargoapi "github.com/akuity/kargo/api/v1alpha1"
+	"github.com/akuity/kargo/internal/api"
+	"github.com/akuity/kargo/internal/api/stubs/rollouts"
+	"github.com/akuity/kargo/internal/expressions"
+)
+
+func (s *server) GetAnalysisRunLogs(
+	ctx context.Context,
+	req *connect.Request[svcv1alpha1.GetAnalysisRunLogsRequest],
+	stream *connect.ServerStream[svcv1alpha1.GetAnalysisRunLogsResponse],
+) error {
+	if !s.cfg.RolloutsIntegrationEnabled {
+		return connect.NewError(
+			connect.CodeUnimplemented,
+			errors.New("Argo Rollouts integration is not enabled"),
+		)
+	}
+
+	namespace := req.Msg.GetNamespace()
+	if err := validateFieldNotEmpty("namespace", namespace); err != nil {
+		return err
+	}
+
+	name := req.Msg.GetName()
+	if err := validateFieldNotEmpty("name", name); err != nil {
+		return err
+	}
+
+	analysisRun, err := rollouts.GetAnalysisRun(
+		ctx,
+		s.client,
+		types.NamespacedName{
+			Namespace: namespace,
+			Name:      name,
+		},
+	)
+	if err != nil {
+		return fmt.Errorf(
+			"error getting AnalysisRun %q in namespace %q: %w",
+			name, namespace, err,
+		)
+	}
+	if analysisRun == nil {
+		return connect.NewError(
+			connect.CodeNotFound,
+			fmt.Errorf("AnalysisRun %q in namespace %q not found", name, namespace),
+		)
+	}
+
+	// Don't stream logs for an AnalysisRun that is not complete.
+	if !analysisRun.Status.Phase.Completed() {
+		return connect.NewError(
+			connect.CodeFailedPrecondition,
+			fmt.Errorf(
+				"AnalysisRun %q in namespace %q is not complete; cannot retrieve logs",
+				name, namespace,
+			),
+		)
+	}
+
+	jobMetricName, jobMetric, err := s.getJobMetric(analysisRun, req.Msg.MetricName)
+	if err != nil {
+		return err
+	}
+
+	containerName, err := s.getContainerName(
+		analysisRun,
+		jobMetricName,
+		jobMetric,
+		req.Msg.ContainerName,
+	)
+	if err != nil {
+		return err
+	}
+
+	jobNamespace, jobName, err := s.getJobNamespaceAndName(analysisRun, jobMetricName)
+	if err != nil {
+		return err
+	}
+
+	stage, err := s.getStageFromAnalysisRun(ctx, analysisRun)
+	if err != nil {
+		return err
+	}
+
+	httpReq, err := s.buildRequest(
+		stage,
+		analysisRun,
+		jobMetricName,
+		jobNamespace,
+		jobName,
+		containerName,
+	)
+	if err != nil {
+		return err
+	}
+
+	httpResp, err := cleanhttp.DefaultClient().Do(httpReq)
+	if err != nil {
+		return fmt.Errorf(
+			"error performing GET request for log url %s: %w",
+			httpReq.URL.String(), err,
+		)
+	}
+	defer httpResp.Body.Close()
+
+	// Logs can be large, so we read them using a buffered reader.
+	reader := bufio.NewReader(httpResp.Body)
+	bufferSize := 4096 // 4 KB chunks
+	buf := make([]byte, bufferSize)
+	for {
```

Comment:
Is it possible for the `ctx` to close before we have finished streaming the whole file?

Reply:
Sure is. Nice catch.